README.md

```diff
@@ -134,7 +134,194 @@ TODO
+> binary version of a message just uses the field's number as the key – the
+> name and declared type for each field can only be determined on the decoding
+> end by referencing the message type's definition (i.e. the .proto file).
+>
+> When a message is encoded, the keys and values are concatenated into a byte
+> stream. When the message is being decoded, the parser needs to be able to
+> skip fields that it doesn't recognize. This way, new fields can be added to a
+> message without breaking old programs that do not know about them. To this
+> end, the "key" for each pair in a wire-format message is actually two values
+> – the field number from your .proto file, plus a wire type that provides just
+> enough information to find the length of the following value.
+>
+> The available wire types are as follows:
+> 
+> Type | Meaning | Used For
+> ---- | ------- | --------
+> 0    | Varint  | int32, int64, uint32, uint64, sint32, sint64, bool, enum
+> 1    | 64-bit  | fixed64, sfixed64, double
+> 2    | Length-delimited | string, bytes, embedded messages, packed repeated fields
+> 3    | Start group | groups (deprecated)
+> 4    | End group | groups (deprecated)
+> 5    | 32-bit  | fixed32, sfixed32, float
+>
+> Each key in the streamed message is a varint with the value (field_number <<
+> 3) | wire_type – in other words, the last three bits of the number store the
+> wire type.
+
+In Wire, a "type" can be encoded by more than 3 bits, due to the introduction
+of a "List" type, which can hold other types recursively.  So collectively,
+lets call these bits the `<type-bytes>`.
+
+In Wire, "varint" means protobuf's "signed varint", and "uvarint" means "varint".
+
+Type | Meaning | Used For
+---- | ------- | --------
+0    | Varint  | bool, byte, [u]int16, and varint-[u]int[64/32]
+1    | 64-bit  | int64, uint64, float64(unsafe)
+2    | Length-delimited | string, bytes, raw?
+3    | Start struct | conceptually, '{'
+4    | End struct | conceptually, '}'; always a single byte, 0x04
+5    | 32-bit  | int32, uint32, float32(unsafe)
+6    | List    | array, slice; followed by 1 or more bytes encoding `<type-bytes>`, then `<uvarint(num-items)>`
+7    | Interface | value starts with `<prefix-bytes>` or `<disfix-bytes>`
+
+Struct fields are encoded in order, and a null/empty/zero field is represented
+by the absence of a field in the encoding, similar to protobuf. In general, the
+total byte-size of a Wire:binary encoded struct cannot be determined in a
+stream until each field's size has been determined by scanning all fields and
+elements recursively.
+
+As in protobuf, each struct field is keyed by a uvarint with the value
+`(field_number << 3) | type`, where `type` is 3 bits long. 
+
+A List is encoded by first writing the type-bytes of the element type, followed
+by the uvarint encoding of the length of the list, followed by the encoding of
+each element.
+
+A list of structs with `n` elements is encoded by the byte `0x03` followed by
```

Comment:
Doesn't this contradict with `Start struct`?

Reply:
The encoding of the value doesn't start with the typ3 byte of itself.  When a value is being encoded as fields in a struct, they are preceded by the field key which includes the typ3 bits.  And, when they are being encoded as items in a slice/array, the typ3 is encoded in the first byte of the list.  So "{" is not *quite* it, because the encoding doesn't start with it, as it is implied.

Reply:
I'm going to rename "Start struct" to "Struct" and "End struct" to "StructTerm" to prevent unnecessary confusion...

Reply:
Also the examples were incorrect.  Examples showed MarshalBinary on a struct starting with 0x03.  That's wrong, it should start with the first field.  Only when a struct is embedded in another struct is it immediately preceded by 0x03.